bindings/communicator_bind.cpp

```diff
@@ -0,0 +1,241 @@
+            typedef void (*f_callback)(void *mesg, int rank, int tag, void *user_data);
+
+            struct callback {
+                f_callback cb;
+                void *user_data = nullptr;
+                callback(f_callback pcb, void *puser_data = nullptr) : cb{pcb}, user_data{puser_data} {}
+                void operator() (communicator_type::message_type message, int rank, int tag) const {
+                    if(cb) cb(&message, rank, tag, user_data);
+                }
+            };
+
+            struct progress_status_type {
+                int num_sends = 0;
+                int num_recvs = 0;
+                int num_cancels = 0;
+                
+                progress_status_type(const ghex::tl::cb::progress_status &ps) :
+                    num_sends{ps.m_num_sends},
+                    num_recvs{ps.m_num_recvs},
+                    num_cancels{ps.m_num_cancels} 
+                {}
+            };
+        }
+    }
+}
+
+extern "C"
+void* ghex_comm_new()
+{
+    return new obj_wrapper(ghex_context->get_communicator());
+}
+
+extern "C"
+int ghex_comm_rank(obj_wrapper *wrapper)
+{
+    return get_object_ptr_unsafe<communicator_type>(wrapper)->rank();
+}
+
+extern "C"
+int ghex_comm_size(obj_wrapper *wrapper)
+{
+    return get_object_ptr_unsafe<communicator_type>(wrapper)->size();
+}
+
+extern "C"
+progress_status_type ghex_comm_progress(obj_wrapper *wrapper)
+{
+    return get_object_ptr_unsafe<communicator_type>(wrapper)->progress();
+}
+
+extern "C"
+void ghex_comm_barrier(obj_wrapper *wrapper, int type)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wrapper);
+    switch(type){
+    case(BarrierThread):
+        ghex_barrier->in_node(*comm);
+        break;
+    case(BarrierRank):
+        ghex_barrier->rank_barrier(*comm);
+        break;
+    default:
+        (*ghex_barrier)(*comm);
+        break;
+    }
+}
+
+/*
+   SEND requests
+ */
+
+extern "C"
+void ghex_comm_post_send(obj_wrapper *wcomm, ghex::tl::cb::any_message *wmessage, int rank, int tag, frequest_type *ffut)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    if(nullptr==wmessage){
+	std::cerr << "ERROR: trying to submit a NULL message in " << __FUNCTION__ << ". Terminating.\n";
+	std::terminate();
+    }
+    auto fut = comm->send(*wmessage, rank, tag);
+    new(ffut->data) decltype(fut)(std::move(fut));
+}
+
+extern "C"
+void ghex_comm_post_send_cb(obj_wrapper *wcomm, ghex::tl::cb::any_message *wmessage, int rank, int tag, f_callback cb, frequest_type *freq, void *user_data)
+{
+    communicator_type *comm = get_object_ptr_unsafe<communicator_type>(wcomm);
+    if(nullptr==wmessage){
+	std::cerr << "ERROR: trying to submit a NULL message in " << __FUNCTION__ << ". Terminating.\n";
+	std::terminate();
+    }
+    auto req = comm->send(*wmessage, rank, tag, callback{cb, user_data});
+    if(!freq) return;
+    new(freq->data) decltype(req)(std::move(req));
```

Comment:
similar question here: how do we control lifetime of these objects allocated in-place? Since they are type erased, we need type injection to properly destruct them at some point.